File: DSA(Geeks for Geeks)/Basics/Queue/04. Implementation of Queue using Linked List.js
// Implementation of Queue using Linked List

/**
 * q = new Queue();
 * q.enque(10);
 * q.enque(20);
 * q.enque(30);
 * console.log(q.deque());
 * q.enque(40);
 * console.log(q.getFront());
 * console.log(q.getRear());
 * console.log(q.size());
 * console.log(q.isEmpty());
 */

/**
 * --> Idea for the Implementation
 * o We need to insert at one end and remove from the other end.
 * 
 *      deque() --> | | | | | | <-- enque()
 *                front       rear
 * 
 * o Enque() and Deque() opeartions both will have O(1) time complexity.
 */

// =================================

class Node {
    constructor(x) {
        this.data = x;
        this.next = null;
    }
}

class Queue {
    constructor() {
        this.front = null;
        this.rear = null;
        this.size = 0;
    }

    enque(x) {
        let temp = new Node(x);
        if(this.front === null) {
            this.front = temp;
        } else {
            this.rear.next = temp;
        }

        this.rear = temp;
        this.size += 1;
    }

    deque() {
        if(this.front === null) {
            return null;
        }
        let res = this.front.data;
        this.front = this.front.next;
        this.size -= 1;
        return res;
    }

    sizeQueue() {
        return this.size
    }

    isEmpty() {
        return this.size === 0;
    }
}

let obj = new Queue();
obj.enque(10);
obj.enque(20);
obj.enque(30);
obj.enque(40);

console.log(obj.sizeQueue());
obj.deque();
console.log(obj.sizeQueue());

File: python/DSA(Geeks for Geeks)/Basics/Queue/04. Implementation of queue using linked list.py
'''
// Implementation of Queue using Linked List

/**
 * q = new Queue();
 * q.enque(10);
 * q.enque(20);
 * q.enque(30);
 * console.log(q.deque());
 * q.enque(40);
 * console.log(q.getFront());
 * console.log(q.getRear());
 * console.log(q.size());
 * console.log(q.isEmpty());
 */

/**
 * --> Idea for the Implementation
 * o We need to insert at one end and remove from the other end.
 * 
 *      deque() --> | | | | | | <-- enque()
 *                front       rear
 * 
 * o Enque() and Deque() opeartions both will have O(1) time complexity.
 */

// =================================
'''

class Node:
    def __init__(self, x):
        self.data = x
        self.next = None

class Queue:
    def __init__(self):
        self.front = None
        self.rear = None
        self.size = 0

    def enque(self, x):
        temp = Node(x)
        if self.front == None:
            self.front = temp
        else:
            self.rear.next = temp

        self.rear = temp
        self.size += 1

    def deque(self):
        if self.front == None:
            return None
        res = self.front.data 
        self.front = self.front.next
        self.size -= 1
        return res
    
    def sizeQueue(self):
        return self.size
    
    def isEmpty(self):
        return self.size == 0
    

obj = Queue()
obj.enque(10)
obj.enque(20)
obj.enque(30)
obj.enque(40)


print(obj.sizeQueue())
obj.deque()
print(obj.sizeQueue())


